### ISSUE TITLE
update or remove `__all__` declarations

### ISSUE THREAD

**Issue by jvanasco** [Pylons/webob#457]
Many of the files in WebOb declare `__all__` for exports.  This can create problems with the current generation of type checkers, as they may not be able to see functions that do not appear in these declarations.  For example, in `webob.cookies`, `parse_cookie` is not declared.

The `__all__` declarations are only used for * imports - which are increasingly unused in exchange for explicit imports.  

I would be happy to generate a PR for either situation, however I would prefer dropping `__all__`.

**Comment by digitalresistor** [comment]
Thanks for raising this. I went through the history and current state of each module before answering. Short version: I'd rather not drop `__all__`, the `parse_cookie` example is working as intended, but you've landed near a real gap — just not the one you cited. Details, since I think the underlying goal is worth getting right.

## `__all__` isn't the lever today — but #477 makes it one

WebOb currently ships no `py.typed` marker and no inline annotations, so nothing in `src/webob/*.py` reaches mypy or pyright. Type checkers resolve `webob` through typeshed's `types-WebOb` stubs (1.8.0.20260716), which carry their own `__all__`. Editing these lists today has no effect on them.

That changes with #477 (tracked by #474), which adds `src/webob/py.typed` and wires it into `setup.py` via `package_data`. Once that lands, the re-export rules become live for real consumers.

Worth separating the two mechanisms, though: `__all__` governs re-export of names a module **imports**, not visibility of names a module **defines**. `parse_cookie` is defined at `src/webob/cookies.py:208`, so it resolves fine either way. Your specific example wouldn't be fixed by dropping or extending `__all__`.

## Why `parse_cookie` isn't exported

It was never intended as public API. It landed in d484fa2 (Sept 2011) as part of *"change req.cookies implementation to parse on demand"* — a helper extracted to serve that refactor. `cookies.__all__` started as `['Cookie']` in 6a83d65 (2010) and has been curated deliberately since. In the fifteen years after, `parse_cookie` has picked up exactly one call site in the library (`src/webob/cookies.py:51`) and no documentation.

I'd also rather not freeze its current signature. It accepts a `str`, latin-1 encodes it (`_parse_cookie`, `src/webob/cookies.py:201-205`), and yields `(bytes, bytes)` pairs — which its sole caller immediately utf-8 decodes both halves of. That asymmetry is an implementation detail. If there's demand for a public cookie-header parser, I'd rather add a properly-shaped `str` → `str` function than promote this one; happy to take that as its own issue.

## The other omissions look deliberate

The criterion `__all__` has been following is *"things a user constructs or calls directly"* — which is why `Cookie` and `Base64Serializer` are listed despite being undocumented, and why these aren't:

- `RequestCookies` — reached, not constructed; the return type of `Request.cookies` (`src/webob/request.py:879`).
- `Morsel` — reached by iterating a `Cookie`.
- `serialize_cookie_date`, `serialize_max_age`, `serialize_samesite`, `cookie_property` — plumbing; their only non-test call sites are the property bindings at `src/webob/cookies.py:278-285`.
- `IfRangeDate`, `FileIter`, `BLOCK_SIZE`, and `response.py`'s `filter_headers` / `iter_file` / `iter_close` / `gzip_app_iter` — same categories.

Dropping `__all__` would promote all of these to public-by-default, which is the wrong direction right as we're trying to commit to a typed surface. It would also break `webob.exc` concretely: `src/webob/exc.py:1325-1334` builds `__all__` dynamically from the exception classes, and `from webob.exc import *` is an idiom we document at `docs/reference.txt:905`. Without it, that star-import starts dumping `re`, `json`, `warnings`, `Request`, and `Response` into the caller's namespace.

## The real gap

There is one place your concern is genuinely correct, and it's in `src/webob/__init__.py` — which #477 does not touch. Six names are imported there for re-export but absent from `__all__`:

`BaseRequest`, `parse_date`, `parse_date_delta`, `serialize_date`, `serialize_date_delta`, `timedelta_to_seconds`

Today that's harmless. Once `py.typed` ships, `from webob import BaseRequest` becomes a pyright `reportPrivateImportUsage` error and a mypy `--no-implicit-reexport` failure for downstream users. Note this won't show up in #477's own CI: `no_implicit_reexport` (implied by the `strict = true` it adds to `pyproject.toml`) flags the *importing* module, and nothing inside WebOb imports these from the top-level package.

So the fix belongs on #477 rather than here — either add the six to `__all__`, or use the redundant-alias form (`from webob.request import BaseRequest as BaseRequest`), which would also let the `# noqa: F401` go. I'll raise it there.

Closing this in favour of #474 / #477, which is where the leverage is — and given you're already reviewing #477, extra eyes on that diff would be very welcome.

**Comment by jvanasco** [comment]
While I'm glad there is some traction on this, I am sad it involves AI.

I think having a properly typed, publicly intended `parse_cookie` is worthwhile.  I use it in some public repos, pyramid_debugtoolbar's session panel uses it too.  IIRC, under python3, I believe it's more likely to be bytes->bytes and not string->string.
